### PR TITLE
[s2sTrainer] test + code cleanup

### PR DESCRIPTION
**Pull request by sshleifer**
- add a 10 second test for Seq2SeqTrainer
- general code cleanup
- pass `data_args` to Seq2SeqTrainer

@patil-suraj

### CODE REVIEW

File: examples/seq2seq/finetune_trainer.py [hunk 1]
```diff
@@ -198,6 +191,19 @@ class DataTrainingArguments:
     eval_beams: Optional[int] = field(default=None, metadata={"help": "# num_beams to use for evaluation."})
 
 
+def freeze_embeds(model: torch.nn.Module):
```
Comment (sshleifer) [inline examples/seq2seq/finetune_trainer.py:194]: Will replace with new FSMT friendly version.

File: examples/seq2seq/finetune_trainer.py [hunk 2]
```diff
@@ -266,17 +260,19 @@ def main():
     use_task_specific_params(model, data_args.task)
 
     # set num_beams for evaluation
-    if data_args.eval_beams is not None:
-        model.config.num_beams = data_args.eval_beams
-    assert model.config.num_beams >= 1, f"got eval_beams={model.config.num_beams}. Need an integer >= 1"
+    if data_args.eval_beams is None:
+        data_args.eval_beams = model.config.num_beams
+    assert data_args.eval_beams >= 1, f"got eval_beams={model.config.num_beams}. Need an integer >= 1"  # TODO: remove?
```
Comment (patil-suraj) [inline examples/seq2seq/finetune_trainer.py:265]: IMO we can remove this `assert`, default is `None` and highly unlikely that someone will set it to 0 or negative

Comment (patil-suraj) [inline examples/seq2seq/finetune_trainer.py:265]: IMO we can remove this assert, default is `None` and highly unlikely that someone will set it to 0 or negative

File: examples/seq2seq/finetune_trainer.py [hunk 3]
```diff
@@ -266,17 +260,19 @@ def main():
...
+    # model.config.max_generate_length = data_args.val_max_target_length
 
     # set decoder_start_token_id for MBart
     if model.config.decoder_start_token_id is None and isinstance(tokenizer, MBartTokenizer):
-        decoder_start_token_id = tokenizer.lang_code_to_id[data_args.tgt_lang]
-        model.config.decoder_start_token_id = decoder_start_token_id
+        assert (
+            data_args.tgt_lang is not None and data_args.src_lang is not None
+        ), "mBart requires --tgt_lang and --src_lang"
+        model.config.decoder_start_token_id = tokenizer.lang_code_to_id[data_args.tgt_lang]
```
Comment (patil-suraj) [inline examples/seq2seq/finetune_trainer.py:275]: Nice!

File: examples/seq2seq/finetune_trainer.py [hunk 4]
```diff
@@ -395,41 +380,34 @@ def freeze_embeds(model: torch.nn.Module):
...
+                test_preds = tokenizer.batch_decode(
+                    test_output.predictions, skip_special_tokens=True, clean_up_tokenization_spaces=True
+                )
                 test_preds = lmap(str.strip, test_preds)
-                output_test_pred_file = os.path.join(training_args.output_dir, "test_generations.txt")
-                with open(output_test_pred_file, "w") as f:
-                    f.write("\n".join(test_preds))
-
+                write_txt_file(test_preds, os.path.join(training_args.output_dir, "test_generations.txt"))
+            # why no eval_results.update(result?)
```
Comment (patil-suraj) [inline examples/seq2seq/finetune_trainer.py:410]: > why no eval_results.update(result?)

initially the metrics keys were same

File: examples/seq2seq/seq2seq_trainer.py [hunk 5]
```diff
@@ -20,6 +20,12 @@
 
 
 class Seq2SeqTrainer(Trainer):
+    def __init__(self, data_args, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.data_args = data_args
+        self.max_gen_length = data_args.val_max_target_length
+        self.pad_token_id = self.model.config.pad_token_id
+
```
Comment (patil-suraj) [inline examples/seq2seq/seq2seq_trainer.py:28]: 👍

File: examples/seq2seq/seq2seq_trainer.py [hunk 6]
```diff
@@ -81,41 +87,32 @@ def prediction_step(
...
                 generated_tokens = self._pad_tensors_to_max_len(
-                    generated_tokens, max_length, model.config.pad_token_id
+                    generated_tokens, self.max_gen_length, self.pad_token_id
                 )
 
             labels_out = inputs.get("labels")
-            outputs = model(**inputs)
-            logits = outputs[1]
-            loss = self._compute_loss(logits, labels_out, model.config.pad_token_id)
+            # Call forward again to get loss # TODO: avoidable?
```
Comment (patil-suraj) [inline examples/seq2seq/seq2seq_trainer.py:105]: probably not avoidable, need `logits` for loss

File: examples/seq2seq/finetune_trainer.py (hunk 2 above)
Comment (sshleifer) [reply examples/seq2seq/finetune_trainer.py:265]: yes. Moving forward try to avoid copying code of mine that obviously makes no sense :)

File: examples/seq2seq/finetune_trainer.py (hunk 4 above)
Comment (sshleifer) [reply examples/seq2seq/finetune_trainer.py:410]: We shouldn't ignore this. I'll fix.

File: examples/seq2seq/seq2seq_trainer.py (hunk 6 above)
Comment (sshleifer) [reply examples/seq2seq/seq2seq_trainer.py:105]: But these are shaped somewhat similarly. If we get them to the same shape we can still call whatever loss function. Not for this PR but would be good to do at some point.
